src/main/java/org/radarbase/authorizer/service/managementportal/CachedManagementPortalClient.java

```diff
@@ -0,0 +1,218 @@
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.radarbase.authorizer.config.ConfigHelper;
+import org.radarbase.authorizer.config.ManagementPortalProperties;
+import org.radarbase.authorizer.service.dto.managementportal.Project;
+import org.radarbase.authorizer.service.dto.managementportal.Subject;
+import org.radarcns.exception.TokenException;
+import org.radarcns.oauth.OAuth2Client;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.stereotype.Service;
+
+@Service
+@ConditionalOnProperty(value = "validator", havingValue = MP_VALIDATOR_PROPERTY_VALUE)
+public class CachedManagementPortalClient implements ManagementPortalClient<Subject, Project> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CachedManagementPortalClient.class);
+  private Set<Subject> subjects;
+  private Set<Project> projects;
+
+  private Duration expiry = Duration.ofHours(1);
+  private Instant lastFetch;
+
+  private OkHttpClient httpClient;
+
+  private OAuth2Client oAuth2Client;
+
+  private ManagementPortalProperties properties;
+
+  private ObjectMapper mapper = new ObjectMapper();
+
+  @Autowired
+  public CachedManagementPortalClient(@Value("${mp-config-file-path}") String mpConfigFilePath)
+      throws MalformedURLException, ConfigurationException {
+    subjects = new HashSet<>();
+    projects = new HashSet<>();
+    lastFetch = Instant.MIN;
+    init(mpConfigFilePath);
+  }
+
+  public CachedManagementPortalClient(@Value("${mp-config-file-path}") String mpConfigFilePath,
+      Duration expiry) throws MalformedURLException, ConfigurationException {
+    this.expiry = expiry;
+    subjects = new HashSet<>();
+    projects = new HashSet<>();
+    lastFetch = Instant.MIN;
+    init(mpConfigFilePath);
+  }
+
+  private void init(String mpConfigFilePath) throws ConfigurationException, MalformedURLException {
+
+    this.httpClient = new OkHttpClient.Builder()
+        .connectTimeout(20, TimeUnit.SECONDS)
+        .writeTimeout(20, TimeUnit.SECONDS)
+        .readTimeout(50, TimeUnit.SECONDS)
+        .build();
+
+    this.mapper = new ObjectMapper()
+        .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+
+    this.properties = loadManagementPortalProperties(mpConfigFilePath);
+    this.oAuth2Client = new OAuth2Client.Builder()
+        .credentials(properties.getOauthClientId(), properties.getOauthClientSecret())
+        .endpoint(new URL(properties.getBaseUrl()), "/oauth/token")
+        .httpClient(httpClient)
+        .build();
+  }
+
+  private ManagementPortalProperties loadManagementPortalProperties(@NotNull String path)
+      throws ConfigurationException {
+    ManagementPortalProperties properties = ConfigHelper
```

Comment:
little confused here with why using YAML models and using property style implementation. I think you can also annotate @NotNull to the class definition. Then it will automatically validate when loading. so no need to do the following check.

Reply:
Do you mean that i should use application.properties for management portal properties instead of a separate file?
Ok good point will add the NotNull checks.

Reply:
Hm. Not exactly. But what you said is even better. read these properties from Application.properties file. So no need to add another config file. and spring application file works with ENV variables for docker, which makes the config easier

Reply:
Done